plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,908 @@
+			return fmt.Errorf("unable to create V2 volume client %v", err)
+		}
+	}
+
+	return nil
+}
+
+// Gather gathers resources from the OpenStack API and accumulates metrics.  This
+// implements the Input interface.
+func (o *OpenStack) Gather(acc telegraf.Accumulator) error {
+	// Gather resources.  Note service harvesting must come first as the other
+	// gatherers are dependant on this information.
+	gatherers := map[string]func(telegraf.Accumulator) error{
+		"projects":      o.gatherProjects,
+		"hypervisors":   o.gatherHypervisors,
+		"flavors":       o.gatherFlavors,
+		"servers":       o.gatherServers,
+		"volumes":       o.gatherVolumes,
+		"storage_pools": o.gatherStoragePools,
+		"subnets":       o.gatherSubnets,
+		"ports":         o.gatherPorts,
+		"networks":      o.gatherNetworks,
+		"aggregates":    o.gatherAggregates,
+		"nova_services": o.gatherNovaServices,
+		"agents":        o.gatherAgents,
+		"stacks":        o.gatherStacks,
+	}
+
+	for _, service := range o.EnabledServices {
+		if service == "services" {
+			o.accumulateServices(acc)
+			continue
+		}
+		gatherer := gatherers[service]
+		if err := gatherer(acc); err != nil {
+			acc.AddError(fmt.Errorf("failed to get resource %q %v", service, err))
+		}
+	}
+
+	if o.ServerDiagnotics && !o.InEnabledServices("servers") {
+		if err := o.gatherServers(acc); err != nil {
+			acc.AddError(fmt.Errorf("failed to get resource server diagnostics %v", err))
+		} else {
+			o.accumulateServerDiagnostics(acc)
+		}
+	}
+
+	return nil
+}
+
+// gatherStacks collects stacks from the OpenStack API.
+func (o *OpenStack) gatherStacks(acc telegraf.Accumulator) error {
+	page, err := stacks.List(o.stack, &stacks.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list stacks %v", err)
+	}
+	stacks, err := stacks.ExtractStacks(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract stacks %v", err)
+	}
+	for _, stack := range stacks {
+		tags := map[string]string{
+			"description": stack.Description,
+			"name":        stack.Name,
+			"stack_tags":  strings.Join(stack.Tags[:], ","),
+		}
+		fields := map[string]interface{}{
+			"status":        stack.Status,
+			"id":            stack.ID,
+			"status_reason": stack.StatusReason,
+			"creation_time": stack.CreationTime.Format("2006-01-02T15:04:05.999999999Z07:00"),
+			"updated_time":  stack.UpdatedTime.Format("2006-01-02T15:04:05.999999999Z07:00"),
+		}
+		acc.AddFields("openstack_stack", fields, tags)
+	}
+
+	return nil
+}
+
+// gatherServices collects services from the OpenStack API.
+func (o *OpenStack) gatherServices() error {
+	page, err := services.List(o.identity, &services.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list services %v", err)
+	}
+	services, err := services.ExtractServices(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract services %v", err)
+	}
+	for _, service := range services {
+		o.services[service.ID] = service
+	}
+
+	return nil
+}
```

Comment:
Can you please move this function up as the first of the `gatherXYZ()` group as it seems special in terms of structure and importance.